src/org/rust/lang/core/psi/util/RustModExtensions.kt

```diff
@@ -0,0 +1,61 @@
+package org.rust.lang.core.psi.util
+
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RustCompositeElement
+import org.rust.lang.core.psi.RustCrateItem
+import org.rust.lang.core.psi.RustModItem
+
+
+private val RustModItem.modDir: PsiDirectory?
+    get() {
+        val parent = parentModInFile ?: return containingFile.parent
+
+        // FIXME(matklad): mod name should be not null, should we rearange our rules for file/mod?
+        val name = this.name ?: return null
+
+        return parent.modDir?.findSubdirectory(name)
+    }
+
+
+val RustCompositeElement.parentModInFile: RustModItem?
+    get() = parentOfType<RustModItem>()
+
+private val RustModItem.isCrateRoot: Boolean
+    get() = parentModInFile == null &&
+        (containingFile.name == "main.rs" || containingFile.name == "lib.rs")
+
+private val RustModItem.ownsDirectory: Boolean
+    get() = parentModInFile != null || // any inline nested module owns a directory
+        containingFile.name == "mod.rs" ||
```

Comment:
Yes, this is intentional. If `foo` is not a directory owner, but `bar` is inlined in `foo`, then `bar` is a directory owner, according to the reference.